File: archives/cpp_gui/menu/compute/compute_model.cpp
#include "compute_model.h"


/* NOTE FOR THE DEVELOPERS:
 * The constructor of a Menu Dialog class deriving from IodeSettingsDialog MUST:
 * 2. call setupUi(this),
 * 3. use the Wrapper classes defined in the wrapper_classes.h header file to wrap the input field items,
 * 4. fill the mapFields member (inherited from IodeSettingsDialog) with the wrapped field items,
 * 5. initialize the inherited className member
 * 6. call the loadSettings() method (inherited from IodeSettingsDialog).
 */


MenuComputeModel::MenuComputeModel(QWidget* parent) : 
    IodeSettingsDialog(parent)
{
    setupUi(this);

	completer = new IodeCompleter(false, false, EQUATIONS, textEdit_equations_list);
	textEdit_equations_list->setCompleter(completer);

    qEquationsList = new WrapperQPlainTextEdit(label_equations_list->text(), *textEdit_equations_list, OPTIONAL_FIELD);
    mapFields["EquationsList"] = qEquationsList;

    className = "MENU_COMPUTE_MODEL";
    loadSettings();
}

MenuComputeModel::~MenuComputeModel()
{
    delete qEquationsList;
    delete completer;
}

void MenuComputeModel::compute()
{
    try
    {
        std::string equations_list = qEquationsList->extractAndVerify().toStdString();

        Simulation sim;

        sim.model_compile(equations_list);

        this->accept();
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
    }
}

File: gui/iode_gui/menu/compute/compute_model.py
from qtpy.QtCore import Slot
from qtpy.QtWidgets import QWidget, QMessageBox

from iode_gui.settings import MixinSettingsDialog
from iode_gui.text_edit.completer import IodeCompleter
from iode_gui.abstract_main_window import AbstractMainWindow
from .ui_compute_model import Ui_MenuComputeModel

import warnings
from iode import IodeType, Simulation


class MenuComputeModel(MixinSettingsDialog):
    def __init__(self, parent: AbstractMainWindow):
        super().__init__(parent)
        self.ui = Ui_MenuComputeModel()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)  

        self.ui.textEdit_equations_list.setup_completer(iode_types=IodeType.EQUATIONS)

        self.load_settings()

    @Slot()
    def compute(self):
        try:
            equations_list: str = self.ui.textEdit_equations_list.toPlainText().strip()

            warnings.simplefilter("error")
            simu = Simulation()
            simu.model_compile(equations_list)         
            warnings.simplefilter("default")

            self.accept()
        except Exception as e:
            self.parent().display_output(f"ERROR -> {str(e)}")
            QMessageBox.warning(self, "WARNING", str(e))
            warnings.simplefilter("default")
